### PR TITLE
Fix S2183 FN: Support unsigned right-shift operator (>>>)

### LINKED ISSUE #6294: Fix S2183 FN: Support unsigned right-shift operator (>>>)
### Description

FN for S2183 (DoNotShiftByZeroOrIntSize).

Using C#11 Unsigned right-shift operator (>>>), the rule is not raising an issue.

### Repro steps

```csharp
        int i;
        ulong j = 0;

        i = i >>> 60; // FN
        i >>>= 40; // FN

        q = q >>> 64; // FN
        q >>>= 70; // FN
```

### Expected behavior

The rule should raise an issue.

### Actual behavior

The rule does not raise an issue.

### PR DESCRIPTION
**Pull request by cristian-ambrosini-sonarsource**
Fixes #6294

### CODE REVIEW

**Comment by gregory-paidis-sonarsource** [comment]
There are a couple of [S109](https://sonarsource.github.io/rspec/#/rspec/S109) raised on lines 105, 109 and  212, but I am not sure how important they are. 

Also, added a small refactor to simplify some ifs, apart from that, looks good.

**Review by gregory-paidis-sonarsource — CHANGES_REQUESTED** [review]

**Review by gregory-paidis-sonarsource — APPROVED** [review]

File: analyzers/src/SonarAnalyzer.CSharp/Rules/DoNotShiftByZeroOrIntSize.cs [hunk 1]
```diff
@@ -97,17 +103,13 @@ protected override void Initialize(SonarAnalysisContext context)
...
-        private static bool ContainsShiftExpressionWithinTwoLines(HashSet<int> linesWithShiftOperations,
-            int lineNumber)
-        {
-            return linesWithShiftOperations.Contains(lineNumber - 2) ||
-                   linesWithShiftOperations.Contains(lineNumber - 1) ||
-                   linesWithShiftOperations.Contains(lineNumber)     ||
-                   linesWithShiftOperations.Contains(lineNumber + 1) ||
-                   linesWithShiftOperations.Contains(lineNumber + 2);
-        }
+        private static bool ContainsShiftExpressionWithinTwoLines(HashSet<int> linesWithShiftOperations, int lineNumber) =>
```
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/DoNotShiftByZeroOrIntSize.cs:107]: Wrong indenting, it should be only 4 spaces.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/DoNotShiftByZeroOrIntSize.cs [hunk 2]
```diff
@@ -35,16 +35,19 @@ namespace SonarAnalyzer.Rules.CSharp
...
-        private const string MessageFormat_UseLargerTypeOrPromote = "Either promote shift target to a larger integer type or shift by {0} instead.";
-        private const string MessageFormat_ShiftTooLarge = "Correct this shift; shift by {0} instead.";
-        private const string MessageFormat_RightShiftTooLarge = "Correct this shift; '{0}' is larger than the type size.";
-        private const string MessageFormat_UselessShift = "Remove this useless shift by {0}.";
+        private const string MessageFormatUseLargerTypeOrPromote = "Either promote shift target to a larger integer type or shift by {0} instead.";
+        private const string MessageFormatShiftTooLarge = "Correct this shift; shift by {0} instead.";
+        private const string MessageFormatRightShiftTooLarge = "Correct this shift; '{0}' is larger than the type size.";
+        private const string MessageFormatUselessShift = "Remove this useless shift by {0}.";
+
+        private const int OneLineShift = 1;
```
[the remark this answers is not in the document]
Comment (cristian-ambrosini-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/DoNotShiftByZeroOrIntSize.cs:43]: And what is your suggestion? Renaming the variable to 'One' and 'Two' or to simply leave the magic numbers in the method as before?

File: analyzers/src/SonarAnalyzer.CSharp/Rules/DoNotShiftByZeroOrIntSize.cs [hunk 3]
```diff
@@ -128,7 +127,7 @@ protected override void Initialize(SonarAnalysisContext context)
                 return new Tuple<Shift, ExpressionSyntax>(Shift.Left, assignmentExpession.Right);
             }
 
-            if (assignmentExpession?.OperatorToken.IsKind(SyntaxKind.GreaterThanGreaterThanEqualsToken) ?? false)
+            if (assignmentExpession?.OperatorToken.IsAnyKind(SyntaxKind.GreaterThanGreaterThanEqualsToken, SyntaxKindEx.GreaterThanGreaterThanGreaterThanEqualsToken) ?? false)
             {
                 return new Tuple<Shift, ExpressionSyntax>(Shift.Right, assignmentExpession.Right);
             }
```
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/DoNotShiftByZeroOrIntSize.cs:133]: NITPICK: In TryGetConstantValue the && operator is at the end of the line however, it should be at the beginning of the next one.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/DoNotShiftByZeroOrIntSize.cs [hunk 4]
```diff
@@ -168,24 +167,22 @@ private static ShiftInstance FindShiftInstance(SyntaxNode node, SemanticModel se
                 return new ShiftInstance(node);
```
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/DoNotShiftByZeroOrIntSize.cs:167]: This and the previous two return statements are identical, I suggest merging the three ifs.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/DoNotShiftByZeroOrIntSize.cs [hunk 5]
```diff
@@ -197,28 +194,28 @@ private static string FindProblemDescription(int typeSizeInBits, int shiftBy, Sh
...
                     "less than " + typeSizeInBits);
             }
 
-            return string.Format(MessageFormat_UseLargerTypeOrPromote, shiftSuggestion);
+            return string.Format(MessageFormatUseLargerTypeOrPromote, shiftSuggestion);
         }
 
-        private class ShiftInstance
+        private sealed class ShiftInstance
         {
```
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/DoNotShiftByZeroOrIntSize.cs:219]: The `public ShiftInstance(SyntaxNode node)` constructor can be turned into an expression-bodied constructor.

**Review by csaba-sagi-sonarsource — CHANGES_REQUESTED** [review]
Nice job, I only left a couple of small comments.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/DoNotShiftByZeroOrIntSize.cs (hunk 4 above)
Comment (csaba-sagi-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/DoNotShiftByZeroOrIntSize.cs:159]: Before your changes, the semantic model was only used if the shift value was a constant, after your last changes we always go to retrieve the type symbol which in theory should have a performance impact. What I had in mind when I suggested merging the three ifs is something like this.

```
            if (!TryGetConstantValue(tuple.Item2, out var shiftByCount)
                || semanticModel.GetTypeInfo(node).ConvertedType is not { } typeSymbol
                || (FindTypeSizeOrDefault(typeSymbol) is var variableBitLength &&  variableBitLength == 0))
```

Comment (cristian-ambrosini-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/DoNotShiftByZeroOrIntSize.cs:159]: Oh I see, that makes more sense, thanks.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/DoNotShiftByZeroOrIntSize.cs [hunk 6]
```diff
@@ -139,8 +138,8 @@ protected override void Initialize(SonarAnalysisContext context)
         private static bool TryGetConstantValue(ExpressionSyntax expression, out int value)
         {
             value = 0;
-            return expression?.RemoveParentheses() is LiteralExpressionSyntax literalExpression &&
-                int.TryParse(literalExpression?.Token.ValueText, out value);
+            return expression?.RemoveParentheses() is LiteralExpressionSyntax literalExpression
```
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/DoNotShiftByZeroOrIntSize.cs:141]: It is odd that the coverage did not improve even if we improved it. On the good side I took a closer look at this return statement and realized that the null conditional access is not needed neither on expression neither on literalExpression.

**Review by csaba-sagi-sonarsource — CHANGES_REQUESTED** [review]
Last change I promise :)

**Review by csaba-sagi-sonarsource — APPROVED** [review]
LGTM!

**Pull request merged**